src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+            reindex_map[other_index] = self.insert_block(other_block);
+        }
+
+        Ok(())
+    }
+
+    /// Creates a minimal sub-chain of `self` that contains all `required_keys`.
+    /// Returns `Error::KeyNotFound` if some of `required_keys` is not present in `self`.
+    ///
+    /// Note: "minimal" means it contains the fewest number of blocks of all such sub-chains.
+    pub fn minimize<'a, I>(&self, required_keys: I) -> Result<Self, Error>
+    where
+        I: IntoIterator<Item = &'a bls::PublicKey>,
+    {
+        // Note: the returned chain is not always strictly minimal. Consider this chain:
+        //
+        //     0->1->2->3
+        //        |
+        //        +->4
+        //
+        // Then calling `minimize([1, 2])` currently returns
+        //
+        //     1->2
+        //     |
+        //     +->4
+        //
+        // Even though the truly minimal chain containing 1 and 2 is just
+        //
+        //     1->2
+        //
+        // This is because 4 lies between 1 and 2 in the underlying `tree` vector and so is
+        // currently included.
+        //
+        // TODO: make this function return the truly minimal chain in all cases.
+
+        let mut min_index = self.len() - 1;
+        let mut max_index = 0;
+
+        for key in required_keys {
+            let index = self.index_of(key).ok_or(Error::KeyNotFound)?;
+            min_index = min_index.min(index);
+            max_index = max_index.max(index);
+        }
+
+        // To account for forks, we also need to include the closest common ancestors of all the
+        // required keys. This is to maintain the invariant that for every key in the chain that is
+        // not the root its parent key is also in the chain.
+        min_index = self.closest_common_ancestor(min_index, max_index);
+
+        let mut chain = Self::new(if min_index == 0 {
+            self.root
+        } else {
+            self.tree[min_index - 1].key
+        });
+
+        for index in min_index..max_index {
+            let block = &self.tree[index];
+
+            chain.tree.push(Block {
+                key: block.key,
+                signature: block.signature.clone(),
+                parent_index: block.parent_index - min_index,
+            })
+        }
+
+        Ok(chain)
+    }
+
+    /// Returns a sub-chain of `self` truncated to the last `count` keys.
+    /// NOTE: a chain must have at least 1 block, so if `count` is 0 it is treated the same as if
+    /// it was 1.
+    pub fn truncate(&self, count: usize) -> Self {
+        let count = count.max(1);
+
+        // Indices of the blocks to include in the result, in reverse order.
+        let mut indices = Vec::with_capacity(count - 1);
+
+        // Index of the currently processed block.
+        let mut index = self.len() - 1;
+
+        // Walk the chain starting from the last key and following the parent links.
```

Comment:
Yes, this is a concern I'm still thinking about how best to solve. The issue with this is that it can make a node that has only part of the chain get stuck if the section switches to another branch of the chain. I have a couple of ideas how to solve this. I think the simplest would be to never truncate the chain to just the last key, but always keep at least a couple of keys (say 10). That would work most of the time, but it's not 100% fix. A simple 100% fix would be to never truncate the chain at all and always use the full chain. There are other options, but they are more complex. I'll probably add one of those fixes in the next PR (the one with the other bugfixes). I might even put up a post on slack or discourse explaining this in more details.

Reply:
Ok, got it. thx for the clarification.